pkg/controller/seed-controller-manager/addon/addon_controller.go

```diff
@@ -435,49 +452,95 @@ func (r *Reconciler) writeAdminKubeconfig(ctx context.Context, log *zap.SugaredL
 	rawManifest := r.combineManifests(rawManifests)
-	manifestFilename, manifestDone, err := r.writeCombinedManifest(log, rawManifest, addon, cluster)
+	manifestFilename, manifestDone, err := r.writeCombinedManifest(log, rawManifest, addon, cluster, addonCluster)
 	if err != nil {
-		return "", "", nil, fmt.Errorf("failed to write all addon resources into a combined manifest file: %v", err)
+		return "", nil, fmt.Errorf("failed to write all addon resources into a combined manifest file: %v", err)
 	}
+	return manifestFilename, manifestDone, err
+}
+
+// seedClusterManifests extracts the addon manifests intended for seed cluster
+func seedClusterManifests(all, user []runtime.RawExtension) (seed []runtime.RawExtension) {
+	m := make(map[string]bool)
+	for _, u := range user {
+		m[string(u.Raw)] = true
+	}
+	for _, x := range all {
+		if !m[string(x.Raw)] {
+			seed = append(seed, x)
+		}
+	}
+	return
+}
 
+// setupManifestInteraction looks up the addon manifests on disk under `addons` directory and templates to a file under `/tmp`
+func (r *Reconciler) setupManifestInteraction(ctx context.Context, log *zap.SugaredLogger, addon *kubermaticv1.Addon, cluster *kubermaticv1.Cluster) manifestInteraction {
+	userManifests, err := r.getAddonManifests(ctx, log, addon, cluster, templateAddonForUserCluster)
+	if err != nil {
+		return manifestInteraction{err: fmt.Errorf("failed to get addon manifests: %v", err)}
+	}
+	allManifests, err := r.getAddonManifests(ctx, log, addon, cluster, templateAddonForSeedCluster)
+	if err != nil {
+		return manifestInteraction{err: fmt.Errorf("failed to get addon manifests: %v", err)}
+	}
+	seedManifests := seedClusterManifests(allManifests, userManifests)
+	userManifestFilename, manifestDone, err := r.writeManifests(userManifests, log, addon, cluster, templateAddonForUserCluster)
+	if err != nil {
+		return manifestInteraction{err: fmt.Errorf("failed to write manifest file for %q: %v", templateAddonForUserCluster, err)}
+	}
+	seedManifestFilename, seedManifestDone, err := r.writeManifests(seedManifests, log, addon, cluster, templateAddonForSeedCluster)
+	if err != nil {
+		return manifestInteraction{
+			done: func() { manifestDone() },
+			err:  fmt.Errorf("failed to write manifest file for %q: %v", templateAddonForSeedCluster, err),
+		}
+	}
 	kubeconfigFilename, kubeconfigDone, err := r.writeAdminKubeconfig(ctx, log, addon, cluster)
 	if err != nil {
-		return "", "", nil, fmt.Errorf("failed to write the admin kubeconfig to the local filesystem: %v", err)
+		return manifestInteraction{
+			done: func() { manifestDone(); seedManifestDone() },
+			err:  fmt.Errorf("failed to write the admin kubeconfig to the local filesystem: %v", err),
+		}
 	}
 
 	done := func() {
 		kubeconfigDone()
 		manifestDone()
+		seedManifestDone()
+	}
+	return manifestInteraction{
+		kubeconfigFilename:   kubeconfigFilename,
+		userManifestFilename: userManifestFilename,
+		seedManifestFilename: seedManifestFilename,
+		done:                 done,
```

Comment:
the count of return parameters was getting a little too large I think so I decided to encapsulate that within a struct for easier consumption